google/cloud/security/common/gcp_type/firewall_rule.py

```diff
@@ -195,40 +213,49 @@ def is_equivalent(self, other):
         """
         return (self.direction == other.direction and
                 self.network == other.network and
-                self.source_tags == other.source_tags and
-                self.target_tags == other.target_tags and
+                self._source_tags == other._source_tags and
+                self._target_tags == other._target_tags and
                 self.source_ranges == other.source_ranges and
                 self.destination_ranges == other.destination_ranges and
                 self.firewall_action.is_equivalent(other.firewall_action))
+    # pylint: enable=protected-access
 
 
 class FirewallAction(object):
     """An association of allowed or denied ports and protocols."""
 
-    def __init__(self, firewall_rule_allowed=None, firewall_rule_denied=None):
+    ALLOW = 'allow'
+    DENY = 'deny'
+    ANY_VALUE = '*'
+
+    def __init__(self, firewall_rules=None, firewall_rule_action='allow'):
         """Initialize.
 
         Args:
-          firewall_rule_allowed (list): A list of dictionaries of allowed ports
-            and protocols.
-          firewall_rule_denied (list): A list of dictionaries of denied ports
+          firewall_rules (list): A list of dictionaries of allowed ports
```

Comment:
I don't like that the same variable can be either a list or a string ('*'). Can this always be a list, but if the list contains '\*' then it is treated as all rules match? See my comment below on using all() to test for that. Make sure  sort_rules can handle this case sanely as well, if it doesn't already.

Commenting here because it won't let me below.

You are missing a comma between 'IPProtocol' and '' in sort_rules, so it ends up returning None instead of an empty string if the key is missing.

Reply:
Done